From de3e0b914b28deaf9034656d8f6e0bdf9ccd7c7f Mon Sep 17 00:00:00 2001
From: Tomohiro Kusumi <kusumi.tomohiro@gmail.com>
Date: Sun, 5 May 2019 08:40:48 +0900
Subject: [PATCH] Linux 5.0 compat: Use totalhigh_pages()
Content-Type: text/plain; charset="utf-8"
Content-Transfer-Encoding: 8bit

Linux kernel commit ca79b0c211af63fa3276f0e3fd7dd9ada2439839
"mm: convert totalram_pages and totalhigh_pages variables to atomic"

replaced `totalhigh_pages` with an inline function `totalhigh_pages()`.
This broke compilation on IA32, etc, as ZoL uses `totalhigh_pages`
on archs with highmem. Confirmed on Fedora 30 (5.0.9-301.fc30.i686).

Reviewed-by: Brian Behlendorf <behlendorf1@llnl.gov>
Signed-off-by: Tomohiro Kusumi <kusumi.tomohiro@gmail.com>
Closes #8677
Closes #8701
Signed-off-by: Colin Ian King <colin.king@canonical.com>
---

Index: zfs-linux-0.7.5/config/kernel-totalhigh_pages.m4
===================================================================
--- /dev/null
+++ zfs-linux-0.7.5/config/kernel-totalhigh_pages.m4
@@ -0,0 +1,19 @@
+dnl #
+dnl # 5.0 API change
+dnl #
+dnl # ca79b0c211af mm: convert totalram_pages and totalhigh_pages variables to atomic
+dnl #
+AC_DEFUN([ZFS_AC_KERNEL_TOTALHIGH_PAGES], [
+	AC_MSG_CHECKING([whether totalhigh_pages() exists])
+	ZFS_LINUX_TRY_COMPILE([
+		#include <linux/highmem.h>
+	],[
+		unsigned long pages __attribute__ ((unused));
+		pages = totalhigh_pages();
+	],[
+		AC_MSG_RESULT(yes)
+		AC_DEFINE(HAVE_TOTALHIGH_PAGES, 1, [totalhigh_pages() exists])
+	],[
+		AC_MSG_RESULT(no)
+	])
+])
Index: zfs-linux-0.7.5/config/kernel.m4
===================================================================
--- zfs-linux-0.7.5.orig/config/kernel.m4
+++ zfs-linux-0.7.5/config/kernel.m4
@@ -132,6 +132,7 @@ AC_DEFUN([ZFS_AC_CONFIG_KERNEL], [
 	ZFS_AC_KERNEL_ACL_HAS_REFCOUNT
 	ZFS_AC_KERNEL_USERNS_CAPABILITIES
 	ZFS_AC_KERNEL_TOTALRAM_PAGES_FUNC
+	ZFS_AC_KERNEL_TOTALHIGH_PAGES
 
 	AS_IF([test "$LINUX_OBJ" != "$LINUX"], [
 		KERNELMAKE_PARAMS="$KERNELMAKE_PARAMS O=$LINUX_OBJ"
Index: zfs-linux-0.7.5/module/zfs/arc.c
===================================================================
--- zfs-linux-0.7.5.orig/module/zfs/arc.c
+++ zfs-linux-0.7.5/module/zfs/arc.c
@@ -3996,7 +3996,7 @@ arc_all_memory(void)
 {
 #ifdef _KERNEL
 #ifdef CONFIG_HIGHMEM
-	return (ptob(zfs_totalram_pages - totalhigh_pages));
+	return (ptob(zfs_totalram_pages - zfs_totalhigh_pages));
 #else
 	return (ptob(zfs_totalram_pages));
 #endif /* CONFIG_HIGHMEM */
